### PR TITLE
Polars: add to_physical transformation

### LINKED ISSUE #2002: Polars: add `to_physical` expression

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #2002

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2004?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2004** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2004?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2003** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2003?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by mccalluc — COMMENTED** [review]
Biggest question is scope. 
- Is this covering all the types it should?
- How would someone just looking at python know there is new functionality after this PR?

I think this has come up before, and I'm not sure exactly how to strike the balance here: we don't need exhaustive docs where Polars does it better, but I also think changes in the functionality available in Python should be reflected in Python docs and tests. A follow-up issue, mentioning this PR might be better than making this wait for more docs.

(And if this is really Rust-only, sorry for the derail.)

File: rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    let cat_domain = active_series
+        .element_domain
+        .as_any()
+        .downcast_ref::<CategoricalDomain>()
+        .ok_or_else(|| {
+            err!(
+                MakeTransformation,
+                "to_physical: only categorical inputs are supported"
+            )
+        })?;
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs:67]: I don't actually need the domains to exist, like I thought I did. Implemented.

File: rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    if !cat_domain.encoding().is_some() {
+        return fallible!(MakeTransformation, "to_physical: to prevent potentially revealing information about row ordering, category ordering must be statically known");
+    }
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs:71]: Two drawbacks to that:
1. Unless we move that to a bit on the transformation struct, the honest-but-curious flag (like other flags) is only enforced for entire functions. So all polars functionality would go behind honest-but-curious.
2. I don't know of a way a user could make the function satisfy stability guarantees by abiding by some contract. That is, it is impossible for the user to behave honestly if they were to disable this check via honest-but-curious.

File: rust/src/transformations/make_stable_expr/expr_to_physical/test.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    let expected = col("data")
+        .cut(vec![-1.0, 1.0], None, false, false)
+        .to_physical();
+
+    let t_clip = expected
+        .clone()
+        .make_stable(lf_domain.clone().select(), SymmetricDistance)?;
+    let actual = t_clip.invoke(&(lf.logical_plan, all()))?.1;
+
+    assert_eq!(expected, actual);
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_to_physical/test.rs:21]: Rewrote all of this.

File: rust/src/transformations/make_stable_expr/expr_to_physical/test.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
+use crate::domains::{LazyFrameDomain, SeriesDomain, SeriesElementDomain};
+use crate::metrics::SymmetricDistance;
+use crate::transformations::make_stable_lazyframe;
+
+use super::*;
+
+fn helper_test_expr_to_physical<
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_to_physical/test.rs:7]:
```suggestion
fn assert_expr_to_physical<
```
Keep it verby?

File: rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                .downcast_ref::<CategoricalDomain>()
+                .ok_or_else(|| {
+                    err!(
+                        MakeTransformation,
+                        "to_physical: expected categorical series domain"
+                    )
+                })?;
+
+            if !cat_domain.encoding().is_some() {
+                return fallible!(MakeTransformation, "to_physical: to prevent potentially revealing information about row ordering, category ordering must be statically known");
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs:82]: If the python interface has an encoding kwarg, should we mention `encoding` in this message to steer the user towards a fix?

**Review by mccalluc — COMMENTED** [review]
Maybe tweak the error message to help the user with the fix?

File: rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs (hunk 5 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_to_physical/mod.rs:82]: This is something an enum dtype would make possible, but I can't steer users there yet.

**Review by mccalluc — APPROVED** [review]
Thanks!

**Pull request merged**